Random deadlock when running `test_shutdown_with_sys_exit_at_pick` on CI

Observed here: https://github.com/joblib/loky/pull/419#issuecomment-2695146641 (platform linux - Python 3.11.11).


```python
tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverShutdown::test_shutdown_with_sys_exit_at_pickle +++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++
~~~~~~~~~~~~~~~ Stack of ExecutorManagerThread (140120246191808) ~~~~~~~~~~~~~~~
  File "/home/runner/miniconda3/envs/testenv/lib/python3.11/threading.py", line 1002, in _bootstrap
    self._bootstrap_inner()
  File "/home/runner/miniconda3/envs/testenv/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "/home/runner/work/loky/loky/loky/process_executor.py", line 637, in run
    self.join_executor_internals()
  File "/home/runner/work/loky/loky/loky/process_executor.py", line 943, in join_executor_internals
    p.join()
  File "/home/runner/miniconda3/envs/testenv/lib/python3.11/multiprocessing/process.py", line 149, in join
    res = self._popen.wait(timeout)
  File "/home/runner/work/loky/loky/loky/backend/popen_loky_posix.py", line 79, in wait
    return self.poll(os.WNOHANG if timeout == 0.0 else 0)
  File "/home/runner/work/loky/loky/loky/backend/popen_loky_posix.py", line 58, in poll
    pid, sts = os.waitpid(self.pid, flag)
~~~~~~~~~~~~~~~~ Stack of Thread-12 (_serve) (140120227317440) ~~~~~~~~~~~~~~~~~
  File "/home/runner/miniconda3/envs/testenv/lib/python3.11/threading.py", line 1002, in _bootstrap
    self._bootstrap_inner()
  File "/home/runner/miniconda3/envs/testenv/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "/home/runner/miniconda3/envs/testenv/lib/python3.11/threading.py", line 982, in run
    self._target(*self._args, **self._kwargs)
  File "/home/runner/miniconda3/envs/testenv/lib/python3.11/multiprocessing/resource_sharer.py", line 138, in _serve
    with self._listener.accept() as conn:
  File "/home/runner/miniconda3/envs/testenv/lib/python3.11/multiprocessing/connection.py", line 480, in accept
    c = self._listener.accept()
  File "/home/runner/miniconda3/envs/testenv/lib/python3.11/multiprocessing/connection.py", line 626, in accept
    s, self._last_accepted = self._socket.accept()
  File "/home/runner/miniconda3/envs/testenv/lib/python3.11/socket.py", line 294, in accept
    fd, addr = self._accept()
+++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++
```

The relevant part of the log seems to be:

<details>

```
2025-03-03T16:23:46.0409687Z tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverShutdown::test_shutdown_with_sys_exit_at_pickle +++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++
2025-03-03T16:23:46.0411073Z ~~~~~~~~~~~~~~~ Stack of ExecutorManagerThread (140120246191808) ~~~~~~~~~~~~~~~
2025-03-03T16:23:46.0411988Z   File "/home/runner/miniconda3/envs/testenv/lib/python3.11/threading.py", line 1002, in _bootstrap
2025-03-03T16:23:46.0412797Z     self._bootstrap_inner()
2025-03-03T16:23:46.0413544Z   File "/home/runner/miniconda3/envs/testenv/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
2025-03-03T16:23:46.0414307Z     self.run()
2025-03-03T16:23:46.0414818Z   File "/home/runner/work/loky/loky/loky/process_executor.py", line 637, in run
2025-03-03T16:23:46.0415494Z     self.join_executor_internals()
2025-03-03T16:23:46.0416203Z   File "/home/runner/work/loky/loky/loky/process_executor.py", line 943, in join_executor_internals
2025-03-03T16:23:46.0417112Z     p.join()
2025-03-03T16:23:46.0417720Z   File "/home/runner/miniconda3/envs/testenv/lib/python3.11/multiprocessing/process.py", line 149, in join
2025-03-03T16:23:46.0418471Z     res = self._popen.wait(timeout)
2025-03-03T16:23:46.0419174Z   File "/home/runner/work/loky/loky/loky/backend/popen_loky_posix.py", line 79, in wait
2025-03-03T16:23:46.0419948Z     return self.poll(os.WNOHANG if timeout == 0.0 else 0)
2025-03-03T16:23:46.0420732Z   File "/home/runner/work/loky/loky/loky/backend/popen_loky_posix.py", line 58, in poll
2025-03-03T16:23:46.0421455Z     pid, sts = os.waitpid(self.pid, flag)
2025-03-03T16:23:46.0422015Z ~~~~~~~~~~~~~~~~ Stack of Thread-12 (_serve) (140120227317440) ~~~~~~~~~~~~~~~~~
2025-03-03T16:23:46.0422960Z   File "/home/runner/miniconda3/envs/testenv/lib/python3.11/threading.py", line 1002, in _bootstrap
2025-03-03T16:23:46.0423733Z     self._bootstrap_inner()
2025-03-03T16:23:46.0424476Z   File "/home/runner/miniconda3/envs/testenv/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
2025-03-03T16:23:46.0425275Z     self.run()
2025-03-03T16:23:46.0425874Z   File "/home/runner/miniconda3/envs/testenv/lib/python3.11/threading.py", line 982, in run
2025-03-03T16:23:46.0426784Z     self._target(*self._args, **self._kwargs)
2025-03-03T16:23:46.0427723Z   File "/home/runner/miniconda3/envs/testenv/lib/python3.11/multiprocessing/resource_sharer.py", line 138, in _serve
2025-03-03T16:23:46.0428631Z     with self._listener.accept() as conn:
2025-03-03T16:23:46.0429495Z   File "/home/runner/miniconda3/envs/testenv/lib/python3.11/multiprocessing/connection.py", line 480, in accept
2025-03-03T16:23:46.0430350Z     c = self._listener.accept()
2025-03-03T16:23:46.0431164Z   File "/home/runner/miniconda3/envs/testenv/lib/python3.11/multiprocessing/connection.py", line 626, in accept
2025-03-03T16:23:46.0432365Z     s, self._last_accepted = self._socket.accept()
2025-03-03T16:23:46.0433177Z   File "/home/runner/miniconda3/envs/testenv/lib/python3.11/socket.py", line 294, in accept
2025-03-03T16:23:46.0433900Z     fd, addr = self._accept()
2025-03-03T16:23:46.0434382Z +++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++
2025-03-03T16:23:47.0221682Z FAILED                                                                   [ 20%]
2025-03-03T16:23:47.0227973Z tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverShutdown::test_shutdown_with_sys_exit_at_pickle ERROR [ 20%]
2025-03-03T16:23:49.0440412Z tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverShutdown::test_interpreter_shutdown ERROR [ 20%]
2025-03-03T16:23:49.0444521Z tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverShutdown::test_interpreter_shutdown ERROR [ 20%]
2025-03-03T16:23:51.0674989Z tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverShutdown::test_hang_issue12364 ERROR [ 21%]
2025-03-03T16:23:51.3210219Z [DEBUG:MainProcess:MainThread] Executor collected: triggering callback for QueueManager wakeup
2025-03-03T16:23:51.4144910Z tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverShutdown::test_hang_issue12364 ERROR [ 21%]
2025-03-03T16:23:51.4145784Z 
2025-03-03T16:23:51.4145990Z ==================================== ERRORS ====================================
2025-03-03T16:23:51.4147114Z _ ERROR at teardown of TestsProcessPoolForkserverShutdown.test_shutdown_with_sys_exit_at_pickle _
2025-03-03T16:23:51.4147835Z 
2025-03-03T16:23:51.4148426Z self = <tests.test_process_executor_forkserver.TestsProcessPoolForkserverShutdown object at 0x7f70472b8f90>
2025-03-03T16:23:51.4150198Z method = <bound method ExecutorShutdownTest.test_shutdown_with_sys_exit_at_pickle of <tests.test_process_executor_forkserver.TestsProcessPoolForkserverShutdown object at 0x7f70472b8f90>>
2025-03-03T16:23:51.4151402Z 
2025-03-03T16:23:51.4151587Z     def teardown_method(self, method):
2025-03-03T16:23:51.4152126Z         # Make sure executor is not broken if it should not be
2025-03-03T16:23:51.4152687Z         executor = getattr(self, "executor", None)
2025-03-03T16:23:51.4153165Z         if executor is not None:
2025-03-03T16:23:51.4153550Z             try:
2025-03-03T16:23:51.4153883Z                 # old pytest markers:
2025-03-03T16:23:51.4154398Z                 expect_broken_pool = hasattr(method, "broken_pool")
2025-03-03T16:23:51.4154928Z                 # new pytest markers:
2025-03-03T16:23:51.4155413Z                 for mark in getattr(method, "pytestmark", []):
2025-03-03T16:23:51.4155938Z                     if mark.name == "broken_pool":
2025-03-03T16:23:51.4156408Z                         expect_broken_pool = True
2025-03-03T16:23:51.4157019Z     
2025-03-03T16:23:51.4157320Z                 if expect_broken_pool:
2025-03-03T16:23:51.4157749Z                     for _ in range(10):
2025-03-03T16:23:51.4158256Z                         # The executor manager thread can take some time to
2025-03-03T16:23:51.4158811Z                         # mark the executor broken.
2025-03-03T16:23:51.4159378Z                         is_actually_broken = executor._flags.broken is not None
2025-03-03T16:23:51.4159939Z                         if is_actually_broken:
2025-03-03T16:23:51.4160374Z                             break
2025-03-03T16:23:51.4160763Z                         sleep(0.1)
2025-03-03T16:23:51.4161145Z                     else:
2025-03-03T16:23:51.4161524Z                         raise AssertionError(
2025-03-03T16:23:51.4161989Z                             "The executor was not flagged broken at the end of "
2025-03-03T16:23:51.4162545Z                             f" {method.__qualname__} as expected."
2025-03-03T16:23:51.4163004Z                         )
2025-03-03T16:23:51.4163338Z                 else:
2025-03-03T16:23:51.4163767Z                     # Check that the executor is not broken right away to avoid
2025-03-03T16:23:51.4164642Z                     # wasting CI time. False negative should be very rare.
2025-03-03T16:23:51.4165330Z                     is_actually_broken = executor._flags.broken is not None
2025-03-03T16:23:51.4165900Z                     assert not is_actually_broken
2025-03-03T16:23:51.4167545Z             finally:
2025-03-03T16:23:51.4168036Z                 # Always shutdown the executor, with SIGKILL if the executor
2025-03-03T16:23:51.4168604Z                 # is actually broken.
2025-03-03T16:23:51.4169083Z                 kill_workers = executor._flags.broken is not None
2025-03-03T16:23:51.4169578Z                 t_start = time.time()
2025-03-03T16:23:51.4170333Z                 executor.shutdown(wait=True, kill_workers=kill_workers)
2025-03-03T16:23:51.4170898Z                 dt = time.time() - t_start
2025-03-03T16:23:51.4171382Z                 assert dt < 10, "Executor took too long to shutdown"
2025-03-03T16:23:51.4171917Z >               _check_subprocesses_number(executor, 0)
2025-03-03T16:23:51.4172240Z 
2025-03-03T16:23:51.4172368Z dt         = 0.00015807151794433594
2025-03-03T16:23:51.4172974Z executor   = <loky.process_executor.ProcessPoolExecutor object at 0x7f70475b9750>
2025-03-03T16:23:51.4173656Z expect_broken_pool = False
2025-03-03T16:23:51.4174051Z is_actually_broken = False
2025-03-03T16:23:51.4174422Z kill_workers = False
2025-03-03T16:23:51.4175646Z method     = <bound method ExecutorShutdownTest.test_shutdown_with_sys_exit_at_pickle of <tests.test_process_executor_forkserver.TestsProcessPoolForkserverShutdown object at 0x7f70472b8f90>>
2025-03-03T16:23:51.4177499Z self       = <tests.test_process_executor_forkserver.TestsProcessPoolForkserverShutdown object at 0x7f70472b8f90>
2025-03-03T16:23:51.4178310Z t_start    = 1741019026.0411627
2025-03-03T16:23:51.4178560Z 
2025-03-03T16:23:51.4178726Z tests/_executor_mixin.py:219: 
2025-03-03T16:23:51.4179177Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2025-03-03T16:23:51.4179793Z tests/_executor_mixin.py:123: in _check_subprocesses_number
2025-03-03T16:23:51.4180315Z     raise e
2025-03-03T16:23:51.4180640Z         children_pids = {2932, 2933, 2934, 2935}
2025-03-03T16:23:51.4181310Z         executor   = <loky.process_executor.ProcessPoolExecutor object at 0x7f70475b9750>
2025-03-03T16:23:51.4181993Z         expected_max_process_number = None
2025-03-03T16:23:51.4182444Z         expected_process_number = 0
2025-03-03T16:23:51.4182857Z         patience   = 10
2025-03-03T16:23:51.4184465Z         pids_cmdlines = [(2932, '/home/runner/miniconda3/envs/testenv/bin/python3.11 -m loky.backend.popen_loky_posix --process-name LokyProce...nner/miniconda3/envs/testenv/bin/python3.11 -m loky.backend.popen_loky_posix --process-name LokyProcess-44 --pipe 27')]
2025-03-03T16:23:51.4185669Z         trial_idx  = 9
2025-03-03T16:23:51.4185879Z         worker_pids = set()
2025-03-03T16:23:51.4186125Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2025-03-03T16:23:51.4186327Z 
2025-03-03T16:23:51.4186756Z executor = <loky.process_executor.ProcessPoolExecutor object at 0x7f70475b9750>
2025-03-03T16:23:51.4187251Z expected_process_number = 0, expected_max_process_number = None, patience = 10
2025-03-03T16:23:51.4187529Z 
2025-03-03T16:23:51.4187616Z     def _check_subprocesses_number(
2025-03-03T16:23:51.4187849Z         executor=None,
2025-03-03T16:23:51.4188067Z         expected_process_number=None,
2025-03-03T16:23:51.4188323Z         expected_max_process_number=None,
2025-03-03T16:23:51.4188558Z         patience=10,
2025-03-03T16:23:51.4188742Z     ):
2025-03-03T16:23:51.4188902Z         if not psutil:
2025-03-03T16:23:51.4189188Z             # psutil is not installed, we cannot check the number of subprocesses
2025-03-03T16:23:51.4189499Z             return
2025-03-03T16:23:51.4189663Z     
2025-03-03T16:23:51.4189842Z         for trial_idx in range(patience):
2025-03-03T16:23:51.4190077Z             try:
2025-03-03T16:23:51.4190297Z                 # Wait for terminating processes to disappear
2025-03-03T16:23:51.4190802Z                 pids_cmdlines = _running_children_pids_with_cmdline(
2025-03-03T16:23:51.4191116Z                     psutil.Process()
2025-03-03T16:23:51.4191337Z                 )
2025-03-03T16:23:51.4191566Z                 children_pids = {pid for pid, _ in pids_cmdlines}
2025-03-03T16:23:51.4191854Z                 if executor is not None:
2025-03-03T16:23:51.4192125Z                     worker_pids = set(executor._processes.keys())
2025-03-03T16:23:51.4192481Z                     # Consistency check: all workers should be in the children list
2025-03-03T16:23:51.4192946Z                     assert worker_pids.issubset(children_pids)
2025-03-03T16:23:51.4193205Z                 else:
2025-03-03T16:23:51.4193468Z                     # Bypass pids checks when executor has been garbage collected
2025-03-03T16:23:51.4193791Z                     worker_pids = children_pids
2025-03-03T16:23:51.4194061Z                 if expected_process_number is not None:
2025-03-03T16:23:51.4194318Z                     assert (
2025-03-03T16:23:51.4194564Z >                       len(children_pids) == expected_process_number
2025-03-03T16:23:51.4194838Z                     ), pids_cmdlines
2025-03-03T16:23:51.4196898Z E                   AssertionError: [(2932, '/home/runner/miniconda3/envs/testenv/bin/python3.11 -m loky.backend.popen_loky_posix --process-name LokyProcess-41 --pipe 24'), (2933, '/home/runner/miniconda3/envs/testenv/bin/python3.11 -m loky.backend.popen_loky_posix --process-name LokyProcess-42 --pipe 25'), (2934, '/home/runner/miniconda3/envs/testenv/bin/python3.11 -m loky.backend.popen_loky_posix --process-name LokyProcess-43 --pipe 26'), (2935, '/home/runner/miniconda3/envs/testenv/bin/python3.11 -m loky.backend.popen_loky_posix --process-name LokyProcess-44 --pipe 27')]
2025-03-03T16:23:51.4198690Z 
2025-03-03T16:23:51.4198778Z children_pids = {2932, 2933, 2934, 2935}
2025-03-03T16:23:51.4199146Z executor   = <loky.process_executor.ProcessPoolExecutor object at 0x7f70475b9750>
2025-03-03T16:23:51.4199529Z expected_max_process_number = None
2025-03-03T16:23:51.4199768Z expected_process_number = 0
2025-03-03T16:23:51.4199976Z patience   = 10
2025-03-03T16:23:51.4200499Z [DEBUG:MainProcess:MainThread] Interpreter shutting down. Waking up 1executor_manager_thread:
2025-03-03T16:23:51.4201379Z [(<_ExecutorManagerThread(ExecutorManagerThread, stopped 140120246191808)>, (<unlocked _thread.lock object at 0x7f70475d4d00>, <loky.process_executor._ThreadWakeup object at 0x7f70474d0050>))]
2025-03-03T16:23:51.4202238Z [DEBUG:MainProcess:MainThread] Interpreter shutting down. Waking up 1executor_manager_thread:
2025-03-03T16:23:51.4203102Z [(<_ExecutorManagerThread(ExecutorManagerThread, stopped 140120246191808)>, (<unlocked _thread.lock object at 0x7f70475d4d00>, <loky.process_executor._ThreadWakeup object at 0x7f70474d0050>))]
2025-03-03T16:23:51.4203966Z [DEBUG:MainProcess:MainThread] Interpreter shutting down. Waking up 1executor_manager_thread:
2025-03-03T16:23:51.4204845Z [(<_ExecutorManagerThread(ExecutorManagerThread, stopped 140120246191808)>, (<unlocked _thread.lock object at 0x7f70475d4d00>, <loky.process_executor._ThreadWakeup object at 0x7f70474d0050>))]
2025-03-03T16:23:51.4205707Z [DEBUG:MainProcess:MainThread] Interpreter shutting down. Waking up 1executor_manager_thread:
2025-03-03T16:23:51.4206716Z [(<_ExecutorManagerThread(ExecutorManagerThread, stopped 140120246191808)>, (<unlocked _thread.lock object at 0x7f70475d4d00>, <loky.process_executor._ThreadWakeup object at 0x7f70474d0050>))]
2025-03-03T16:23:51.4207564Z [DEBUG:MainProcess:MainThread] Interpreter shutting down. Waking up 1executor_manager_thread:
2025-03-03T16:23:51.4208414Z [(<_ExecutorManagerThread(ExecutorManagerThread, stopped 140120246191808)>, (<unlocked _thread.lock object at 0x7f70475d4d00>, <loky.process_executor._ThreadWakeup object at 0x7f70474d0050>))]
2025-03-03T16:23:51.4209261Z [DEBUG:MainProcess:MainThread] Interpreter shutting down. Waking up 1executor_manager_thread:
2025-03-03T16:23:51.4210243Z [(<_ExecutorManagerThread(ExecutorManagerThread, stopped 140120246191808)>, (<unlocked _thread.lock object at 0x7f70475d4d00>, <loky.process_executor._ThreadWakeup object at 0x7f70474d0050>))]
2025-03-03T16:23:51.4211100Z [DEBUG:MainProcess:MainThread] Interpreter shutting down. Waking up 1executor_manager_thread:
2025-03-03T16:23:51.4211949Z [(<_ExecutorManagerThread(ExecutorManagerThread, stopped 140120246191808)>, (<unlocked _thread.lock object at 0x7f70475d4d00>, <loky.process_executor._ThreadWakeup object at 0x7f70474d0050>))]
2025-03-03T16:23:51.4212676Z [INFO:MainProcess:MainThread] process shutting down
2025-03-03T16:23:51.4213186Z [DEBUG:MainProcess:MainThread] running all "atexit" finalizers with priority >= 0
2025-03-03T16:23:51.4213654Z [INFO:MainProcess:MainThread] calling join() for process LokyProcess-44
2025-03-03T16:23:51.4214717Z pids_cmdlines = [(2932, '/home/runner/miniconda3/envs/testenv/bin/python3.11 -m loky.backend.popen_loky_posix --process-name LokyProce...nner/miniconda3/envs/testenv/bin/python3.11 -m loky.backend.popen_loky_posix --process-name LokyProcess-44 --pipe 27')]
2025-03-03T16:23:51.4215648Z trial_idx  = 9
2025-03-03T16:23:51.4215824Z worker_pids = set()
2025-03-03T16:23:51.4215936Z 
2025-03-03T16:23:51.4216041Z tests/_executor_mixin.py:102: AssertionError
2025-03-03T16:23:51.4216373Z ---------------------------- Captured stderr setup -----------------------------
2025-03-03T16:23:51.4216941Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120880984064
2025-03-03T16:23:51.4217378Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120880979968
2025-03-03T16:23:51.4217814Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120856637440
2025-03-03T16:23:51.4218240Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120846254080
2025-03-03T16:23:51.4218663Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120846249984
2025-03-03T16:23:51.4219092Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120804052992
2025-03-03T16:23:51.4219498Z [DEBUG:MainProcess:MainThread] ProcessPoolExecutor is setup
2025-03-03T16:23:51.4219895Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120804048896
2025-03-03T16:23:51.4220314Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120804044800
2025-03-03T16:23:51.4220736Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120804040704
2025-03-03T16:23:51.4221158Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120804036608
2025-03-03T16:23:51.4221574Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120804032512
2025-03-03T16:23:51.4222417Z [DEBUG:MainProcess:MainThread] Adjusted process count to 5: [('ForkServerProcess-36', 2925), ('ForkServerProcess-37', 2926), ('ForkServerProcess-38', 2927), ('ForkServerProcess-39', 2928), ('ForkServerProcess-40', 2929)]
2025-03-03T16:23:51.4223243Z [DEBUG:MainProcess:MainThread] _start_executor_manager_thread called
2025-03-03T16:23:51.4223671Z [DEBUG/ForkServerProcess-36] Using 'cloudpickle' for serialization.
2025-03-03T16:23:51.4224084Z [DEBUG:MainProcess:ExecutorManagerThread] Queue._start_thread()
2025-03-03T16:23:51.4224492Z [DEBUG:MainProcess:ExecutorManagerThread] doing self._thread.start()
2025-03-03T16:23:51.4224941Z [DEBUG/ForkServerProcess-37] Using 'cloudpickle' for serialization.
2025-03-03T16:23:51.4225377Z [DEBUG:MainProcess:QueueFeederThread] starting thread to feed data to pipe
2025-03-03T16:23:51.4225824Z [DEBUG:MainProcess:ExecutorManagerThread] ... done self._thread.start()
2025-03-03T16:23:51.4226247Z [DEBUG/ForkServerProcess-38] Using 'cloudpickle' for serialization.
2025-03-03T16:23:51.4226861Z [DEBUG/ForkServerProcess-36] recreated blocker with handle 140120880979968
2025-03-03T16:23:51.4227304Z [DEBUG/ForkServerProcess-36] recreated blocker with handle 140120856637440
2025-03-03T16:23:51.4227737Z [DEBUG/ForkServerProcess-36] recreated blocker with handle 140120846254080
2025-03-03T16:23:51.4228180Z [DEBUG/ForkServerProcess-36] recreated blocker with handle 140120846249984
2025-03-03T16:23:51.4228753Z [DEBUG/ForkServerProcess-36] recreated blocker with handle 140120804052992
2025-03-03T16:23:51.4229197Z [DEBUG/ForkServerProcess-36] recreated blocker with handle 140120893419520
2025-03-03T16:23:51.4229637Z [DEBUG/ForkServerProcess-36] recreated blocker with handle 140120893415424
2025-03-03T16:23:51.4230075Z [DEBUG/ForkServerProcess-36] recreated blocker with handle 140120893411328
2025-03-03T16:23:51.4230509Z [DEBUG/ForkServerProcess-36] recreated blocker with handle 140120889061376
2025-03-03T16:23:51.4230947Z [DEBUG/ForkServerProcess-36] recreated blocker with handle 140120888881152
2025-03-03T16:23:51.4231512Z [DEBUG/ForkServerProcess-36] recreated blocker with handle 140120880984064
2025-03-03T16:23:51.4231949Z [DEBUG/ForkServerProcess-36] recreated blocker with handle 140120804048896
2025-03-03T16:23:51.4232366Z [INFO/ForkServerProcess-36] child process calling self.run()
2025-03-03T16:23:51.4232779Z [DEBUG/ForkServerProcess-37] recreated blocker with handle 140120880979968
2025-03-03T16:23:51.4233192Z [DEBUG/ForkServerProcess-36] Worker started with timeout=None
2025-03-03T16:23:51.4233603Z [DEBUG/ForkServerProcess-37] recreated blocker with handle 140120856637440
2025-03-03T16:23:51.4234102Z [DEBUG/ForkServerProcess-36] Enabling faulthandler to report tracebacks on worker crashes.
2025-03-03T16:23:51.4234611Z [DEBUG/ForkServerProcess-37] recreated blocker with handle 140120846254080
2025-03-03T16:23:51.4235050Z [DEBUG/ForkServerProcess-37] recreated blocker with handle 140120846249984
2025-03-03T16:23:51.4235487Z [DEBUG/ForkServerProcess-37] recreated blocker with handle 140120804052992
2025-03-03T16:23:51.4235924Z [DEBUG/ForkServerProcess-38] recreated blocker with handle 140120880979968
2025-03-03T16:23:51.4236361Z [DEBUG/ForkServerProcess-37] recreated blocker with handle 140120893419520
2025-03-03T16:23:51.4236991Z [DEBUG/ForkServerProcess-38] recreated blocker with handle 140120856637440
2025-03-03T16:23:51.4237435Z [DEBUG/ForkServerProcess-37] recreated blocker with handle 140120893415424
2025-03-03T16:23:51.4237870Z [DEBUG/ForkServerProcess-38] recreated blocker with handle 140120846254080
2025-03-03T16:23:51.4238301Z [DEBUG/ForkServerProcess-37] recreated blocker with handle 140120893411328
2025-03-03T16:23:51.4238738Z [DEBUG/ForkServerProcess-37] recreated blocker with handle 140120889061376
2025-03-03T16:23:51.4239174Z [DEBUG/ForkServerProcess-38] recreated blocker with handle 140120846249984
2025-03-03T16:23:51.4239614Z [DEBUG/ForkServerProcess-37] recreated blocker with handle 140120888881152
2025-03-03T16:23:51.4240061Z [DEBUG/ForkServerProcess-38] recreated blocker with handle 140120804052992
2025-03-03T16:23:51.4240500Z [DEBUG/ForkServerProcess-37] recreated blocker with handle 140120880984064
2025-03-03T16:23:51.4240933Z [DEBUG/ForkServerProcess-37] recreated blocker with handle 140120804044800
2025-03-03T16:23:51.4241343Z [INFO/ForkServerProcess-37] child process calling self.run()
2025-03-03T16:23:51.4241721Z [DEBUG/ForkServerProcess-37] Worker started with timeout=None
2025-03-03T16:23:51.4242191Z [DEBUG/ForkServerProcess-37] Enabling faulthandler to report tracebacks on worker crashes.
2025-03-03T16:23:51.4242679Z [DEBUG/ForkServerProcess-39] Using 'cloudpickle' for serialization.
2025-03-03T16:23:51.4243108Z [DEBUG/ForkServerProcess-38] recreated blocker with handle 140120893419520
2025-03-03T16:23:51.4243540Z [DEBUG/ForkServerProcess-38] recreated blocker with handle 140120893415424
2025-03-03T16:23:51.4243980Z [DEBUG/ForkServerProcess-38] recreated blocker with handle 140120893411328
2025-03-03T16:23:51.4244419Z [DEBUG/ForkServerProcess-38] recreated blocker with handle 140120889061376
2025-03-03T16:23:51.4244856Z [DEBUG/ForkServerProcess-38] recreated blocker with handle 140120888881152
2025-03-03T16:23:51.4245295Z [DEBUG/ForkServerProcess-38] recreated blocker with handle 140120880984064
2025-03-03T16:23:51.4245729Z [DEBUG/ForkServerProcess-38] recreated blocker with handle 140120804040704
2025-03-03T16:23:51.4246246Z [INFO/ForkServerProcess-38] child process calling self.run()
2025-03-03T16:23:51.4246798Z [DEBUG/ForkServerProcess-38] Worker started with timeout=None
2025-03-03T16:23:51.4247262Z [DEBUG/ForkServerProcess-38] Enabling faulthandler to report tracebacks on worker crashes.
2025-03-03T16:23:51.4247748Z [DEBUG/ForkServerProcess-40] Using 'cloudpickle' for serialization.
2025-03-03T16:23:51.4248146Z ----------------------------- Captured stderr call -----------------------------
2025-03-03T16:23:51.4248712Z [DEBUG:MainProcess:MainThread] shutting down executor <loky.process_executor.ProcessPoolExecutor object at 0x7f70475b9750>
2025-03-03T16:23:51.4249518Z [DEBUG:MainProcess:ExecutorManagerThread] releasing worker exit lock on ForkServerProcess-36
2025-03-03T16:23:51.4250104Z [DEBUG:MainProcess:ExecutorManagerThread] releasing worker exit lock on ForkServerProcess-37
2025-03-03T16:23:51.4250679Z [DEBUG:MainProcess:ExecutorManagerThread] releasing worker exit lock on ForkServerProcess-38
2025-03-03T16:23:51.4251252Z [DEBUG:MainProcess:ExecutorManagerThread] releasing worker exit lock on ForkServerProcess-39
2025-03-03T16:23:51.4251821Z [DEBUG:MainProcess:ExecutorManagerThread] releasing worker exit lock on ForkServerProcess-40
2025-03-03T16:23:51.4252314Z [DEBUG:MainProcess:ExecutorManagerThread] found 5 processes to stop
2025-03-03T16:23:51.4252753Z [DEBUG:MainProcess:ExecutorManagerThread] sent 5 sentinels to the call queue
2025-03-03T16:23:51.4253175Z [DEBUG:MainProcess:ExecutorManagerThread] closing call_queue
2025-03-03T16:23:51.4253574Z [DEBUG:MainProcess:ExecutorManagerThread] telling queue thread to quit
2025-03-03T16:23:51.4253977Z [DEBUG:MainProcess:ExecutorManagerThread] Queue.join_thread()
2025-03-03T16:23:51.4254403Z [DEBUG:MainProcess:QueueFeederThread] feeder thread got sentinel -- exiting
2025-03-03T16:23:51.4254827Z [DEBUG:MainProcess:ExecutorManagerThread] closing result_queue
2025-03-03T16:23:51.4255212Z [DEBUG:MainProcess:ExecutorManagerThread] closing thread_wakeup
2025-03-03T16:23:51.4255595Z [DEBUG:MainProcess:ExecutorManagerThread] joining 5 processes
2025-03-03T16:23:51.4256069Z [DEBUG:MainProcess:ExecutorManagerThread] joining process ForkServerProcess-40 with pid 2929
2025-03-03T16:23:51.4256698Z [INFO/ForkServerProcess-37] Shutting down worker on sentinel
2025-03-03T16:23:51.4257065Z [INFO/ForkServerProcess-38] Shutting down worker on sentinel
2025-03-03T16:23:51.4257379Z [DEBUG/ForkServerProcess-37] Exited cleanly
2025-03-03T16:23:51.4257669Z [INFO/ForkServerProcess-37] process shutting down
2025-03-03T16:23:51.4258068Z [DEBUG/ForkServerProcess-37] running all "atexit" finalizers with priority >= 0
2025-03-03T16:23:51.4258465Z [DEBUG/ForkServerProcess-38] Exited cleanly
2025-03-03T16:23:51.4258816Z [DEBUG/ForkServerProcess-37] running the remaining "atexit" finalizers
2025-03-03T16:23:51.4259185Z [INFO/ForkServerProcess-38] process shutting down
2025-03-03T16:23:51.4259579Z [DEBUG/ForkServerProcess-38] running all "atexit" finalizers with priority >= 0
2025-03-03T16:23:51.4260015Z [INFO/ForkServerProcess-37] process exiting with exitcode 0
2025-03-03T16:23:51.4260413Z [DEBUG/ForkServerProcess-38] running the remaining "atexit" finalizers
2025-03-03T16:23:51.4260808Z [INFO/ForkServerProcess-38] process exiting with exitcode 0
2025-03-03T16:23:51.4261196Z [DEBUG/ForkServerProcess-36] psutil return memory size: 28295168
2025-03-03T16:23:51.4261587Z [INFO/ForkServerProcess-36] Shutting down worker on sentinel
2025-03-03T16:23:51.4261906Z [DEBUG/ForkServerProcess-36] Exited cleanly
2025-03-03T16:23:51.4262185Z [INFO/ForkServerProcess-36] process shutting down
2025-03-03T16:23:51.4262575Z [DEBUG/ForkServerProcess-36] running all "atexit" finalizers with priority >= 0
2025-03-03T16:23:51.4263043Z [DEBUG/ForkServerProcess-36] running the remaining "atexit" finalizers
2025-03-03T16:23:51.4263445Z [INFO/ForkServerProcess-36] process exiting with exitcode 0
2025-03-03T16:23:51.4263851Z [DEBUG/ForkServerProcess-40] recreated blocker with handle 140120880979968
2025-03-03T16:23:51.4264419Z [DEBUG/ForkServerProcess-39] recreated blocker with handle 140120880979968
2025-03-03T16:23:51.4264866Z [DEBUG/ForkServerProcess-39] recreated blocker with handle 140120856637440
2025-03-03T16:23:51.4265309Z [DEBUG/ForkServerProcess-40] recreated blocker with handle 140120856637440
2025-03-03T16:23:51.4265748Z [DEBUG/ForkServerProcess-40] recreated blocker with handle 140120846254080
2025-03-03T16:23:51.4266178Z [DEBUG/ForkServerProcess-39] recreated blocker with handle 140120846254080
2025-03-03T16:23:51.4266717Z [DEBUG/ForkServerProcess-39] recreated blocker with handle 140120846249984
2025-03-03T16:23:51.4267163Z [DEBUG/ForkServerProcess-40] recreated blocker with handle 140120846249984
2025-03-03T16:23:51.4267719Z [DEBUG/ForkServerProcess-39] recreated blocker with handle 140120804052992
2025-03-03T16:23:51.4268159Z [DEBUG/ForkServerProcess-40] recreated blocker with handle 140120804052992
2025-03-03T16:23:51.4268598Z [DEBUG/ForkServerProcess-39] recreated blocker with handle 140120893419520
2025-03-03T16:23:51.4269034Z [DEBUG/ForkServerProcess-40] recreated blocker with handle 140120893419520
2025-03-03T16:23:51.4269476Z [DEBUG/ForkServerProcess-40] recreated blocker with handle 140120893415424
2025-03-03T16:23:51.4269914Z [DEBUG/ForkServerProcess-39] recreated blocker with handle 140120893415424
2025-03-03T16:23:51.4270349Z [DEBUG/ForkServerProcess-39] recreated blocker with handle 140120893411328
2025-03-03T16:23:51.4270787Z [DEBUG/ForkServerProcess-40] recreated blocker with handle 140120893411328
2025-03-03T16:23:51.4271225Z [DEBUG/ForkServerProcess-40] recreated blocker with handle 140120889061376
2025-03-03T16:23:51.4271659Z [DEBUG/ForkServerProcess-39] recreated blocker with handle 140120889061376
2025-03-03T16:23:51.4272101Z [DEBUG/ForkServerProcess-39] recreated blocker with handle 140120888881152
2025-03-03T16:23:51.4272543Z [DEBUG/ForkServerProcess-40] recreated blocker with handle 140120888881152
2025-03-03T16:23:51.4272981Z [DEBUG/ForkServerProcess-40] recreated blocker with handle 140120880984064
2025-03-03T16:23:51.4273432Z [DEBUG/ForkServerProcess-39] recreated blocker with handle 140120880984064
2025-03-03T16:23:51.4273864Z [DEBUG/ForkServerProcess-40] recreated blocker with handle 140120804032512
2025-03-03T16:23:51.4274302Z [DEBUG/ForkServerProcess-39] recreated blocker with handle 140120804036608
2025-03-03T16:23:51.4274719Z [INFO/ForkServerProcess-40] child process calling self.run()
2025-03-03T16:23:51.4275095Z [INFO/ForkServerProcess-39] child process calling self.run()
2025-03-03T16:23:51.4275468Z [DEBUG/ForkServerProcess-39] Worker started with timeout=None
2025-03-03T16:23:51.4275843Z [DEBUG/ForkServerProcess-40] Worker started with timeout=None
2025-03-03T16:23:51.4276420Z [DEBUG/ForkServerProcess-39] Enabling faulthandler to report tracebacks on worker crashes.
2025-03-03T16:23:51.4277109Z [DEBUG/ForkServerProcess-40] Enabling faulthandler to report tracebacks on worker crashes.
2025-03-03T16:23:51.4277580Z [INFO/ForkServerProcess-39] Shutting down worker on sentinel
2025-03-03T16:23:51.4277951Z [INFO/ForkServerProcess-40] Shutting down worker on sentinel
2025-03-03T16:23:51.4278270Z [DEBUG/ForkServerProcess-39] Exited cleanly
2025-03-03T16:23:51.4278539Z [DEBUG/ForkServerProcess-40] Exited cleanly
2025-03-03T16:23:51.4278818Z [INFO/ForkServerProcess-39] process shutting down
2025-03-03T16:23:51.4279118Z [INFO/ForkServerProcess-40] process shutting down
2025-03-03T16:23:51.4279505Z [DEBUG/ForkServerProcess-39] running all "atexit" finalizers with priority >= 0
2025-03-03T16:23:51.4280016Z [DEBUG/ForkServerProcess-40] running all "atexit" finalizers with priority >= 0
2025-03-03T16:23:51.4280482Z [DEBUG/ForkServerProcess-39] running the remaining "atexit" finalizers
2025-03-03T16:23:51.4280918Z [DEBUG/ForkServerProcess-40] running the remaining "atexit" finalizers
2025-03-03T16:23:51.4281323Z [INFO/ForkServerProcess-39] process exiting with exitcode 0
2025-03-03T16:23:51.4281688Z [INFO/ForkServerProcess-40] process exiting with exitcode 0
2025-03-03T16:23:51.4282274Z [DEBUG:MainProcess:ExecutorManagerThread] joining process ForkServerProcess-39 with pid 2928
2025-03-03T16:23:51.4282969Z [DEBUG:MainProcess:ExecutorManagerThread] joining process ForkServerProcess-38 with pid 2927
2025-03-03T16:23:51.4283549Z [DEBUG:MainProcess:ExecutorManagerThread] joining process ForkServerProcess-37 with pid 2926
2025-03-03T16:23:51.4284114Z [DEBUG:MainProcess:ExecutorManagerThread] joining process ForkServerProcess-36 with pid 2925
2025-03-03T16:23:51.4284704Z [DEBUG:MainProcess:ExecutorManagerThread] executor management thread clean shutdown of 5 workers
2025-03-03T16:23:51.4285267Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120880984064 and name "None"
2025-03-03T16:23:51.4285899Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120880979968 and name "None"
2025-03-03T16:23:51.4286420Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120856637440 and name "None"
2025-03-03T16:23:51.4287041Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120846254080 and name "None"
2025-03-03T16:23:51.4287555Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120846249984 and name "None"
2025-03-03T16:23:51.4288068Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120804052992 and name "None"
2025-03-03T16:23:51.4288513Z [DEBUG:MainProcess:MainThread] ProcessPoolExecutor is setup
2025-03-03T16:23:51.4288945Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120804048896 and name "None"
2025-03-03T16:23:51.4289417Z [DEBUG:MainProcess:MainThread] launched python with pid 2932 and cmd:
2025-03-03T16:23:51.4290095Z ['/home/runner/miniconda3/envs/testenv/bin/python3.11', '-m', 'loky.backend.popen_loky_posix', '--process-name', 'LokyProcess-41', '--pipe', '24']
2025-03-03T16:23:51.4290804Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120804044800 and name "None"
2025-03-03T16:23:51.4291277Z [DEBUG:MainProcess:MainThread] launched python with pid 2933 and cmd:
2025-03-03T16:23:51.4291936Z ['/home/runner/miniconda3/envs/testenv/bin/python3.11', '-m', 'loky.backend.popen_loky_posix', '--process-name', 'LokyProcess-42', '--pipe', '25']
2025-03-03T16:23:51.4292640Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120804040704 and name "None"
2025-03-03T16:23:51.4293108Z [DEBUG:MainProcess:MainThread] launched python with pid 2934 and cmd:
2025-03-03T16:23:51.4293756Z ['/home/runner/miniconda3/envs/testenv/bin/python3.11', '-m', 'loky.backend.popen_loky_posix', '--process-name', 'LokyProcess-43', '--pipe', '26']
2025-03-03T16:23:51.4294450Z [DEBUG:MainProcess:MainThread] created semlock with handle 140120804036608 and name "None"
2025-03-03T16:23:51.4294910Z [DEBUG:MainProcess:MainThread] launched python with pid 2935 and cmd:
2025-03-03T16:23:51.4295564Z ['/home/runner/miniconda3/envs/testenv/bin/python3.11', '-m', 'loky.backend.popen_loky_posix', '--process-name', 'LokyProcess-44', '--pipe', '27']
2025-03-03T16:23:51.4296561Z [DEBUG:MainProcess:MainThread] Adjusted process count to 4: [('LokyProcess-41', 2932), ('LokyProcess-42', 2933), ('LokyProcess-43', 2934), ('LokyProcess-44', 2935)]
2025-03-03T16:23:51.4297227Z [DEBUG:MainProcess:MainThread] _start_executor_manager_thread called
2025-03-03T16:23:51.4297633Z [DEBUG:MainProcess:ExecutorManagerThread] Queue._start_thread()
2025-03-03T16:23:51.4298041Z [DEBUG:MainProcess:ExecutorManagerThread] doing self._thread.start()
2025-03-03T16:23:51.4298477Z [DEBUG:MainProcess:QueueFeederThread] starting thread to feed data to pipe
2025-03-03T16:23:51.4298925Z [DEBUG:MainProcess:ExecutorManagerThread] ... done self._thread.start()
2025-03-03T16:23:51.4299530Z [DEBUG:MainProcess:MainThread] shutting down executor <loky.process_executor.ProcessPoolExecutor object at 0x7f704749d090>
2025-03-03T16:23:51.4300189Z [DEBUG:MainProcess:ExecutorManagerThread] releasing worker exit lock on LokyProcess-41
2025-03-03T16:23:51.4300728Z [DEBUG:MainProcess:ExecutorManagerThread] releasing worker exit lock on LokyProcess-42
2025-03-03T16:23:51.4301258Z [DEBUG:MainProcess:ExecutorManagerThread] releasing worker exit lock on LokyProcess-43
2025-03-03T16:23:51.4301894Z [DEBUG:MainProcess:ExecutorManagerThread] releasing worker exit lock on LokyProcess-44
2025-03-03T16:23:51.4302378Z [DEBUG:MainProcess:ExecutorManagerThread] found 4 processes to stop
2025-03-03T16:23:51.4302823Z [DEBUG:MainProcess:ExecutorManagerThread] sent 4 sentinels to the call queue
2025-03-03T16:23:51.4303256Z [DEBUG:MainProcess:ExecutorManagerThread] closing call_queue
2025-03-03T16:23:51.4303661Z [DEBUG:MainProcess:ExecutorManagerThread] telling queue thread to quit
2025-03-03T16:23:51.4304063Z [DEBUG:MainProcess:ExecutorManagerThread] Queue.join_thread()
2025-03-03T16:23:51.4304615Z [DEBUG:MainProcess:ExecutorManagerThread] closing result_queue
2025-03-03T16:23:51.4305000Z [DEBUG:MainProcess:ExecutorManagerThread] closing thread_wakeup
2025-03-03T16:23:51.4305382Z [DEBUG:MainProcess:ExecutorManagerThread] joining 4 processes
2025-03-03T16:23:51.4305835Z [DEBUG:MainProcess:ExecutorManagerThread] joining process LokyProcess-44 with pid 2935
2025-03-03T16:23:51.4306436Z [DEBUG:LokyProcess-41:MainThread] recreated blocker with handle 140120880979968 and name "/loky-2214-vz_11jyo"
2025-03-03T16:23:51.4307179Z [DEBUG:LokyProcess-41:MainThread] recreated blocker with handle 140120856637440 and name "/loky-2214-jb7x32rn"
2025-03-03T16:23:51.4307819Z [DEBUG:LokyProcess-41:MainThread] recreated blocker with handle 140120846254080 and name "/loky-2214-9n25pawk"
2025-03-03T16:23:51.4308461Z [DEBUG:LokyProcess-41:MainThread] recreated blocker with handle 140120846249984 and name "/loky-2214-e48w3t5b"
2025-03-03T16:23:51.4309092Z [DEBUG:LokyProcess-41:MainThread] recreated blocker with handle 140120804052992 and name "/loky-2214-gut13v_x"
2025-03-03T16:23:51.4309729Z [DEBUG:LokyProcess-41:MainThread] recreated blocker with handle 140120880984064 and name "/loky-2214-z9o76fg3"
2025-03-03T16:23:51.4310364Z [DEBUG:LokyProcess-41:MainThread] recreated blocker with handle 140120804048896 and name "/loky-2214-j7g5yj28"
2025-03-03T16:23:51.4310885Z [INFO:LokyProcess-41:MainThread] child process calling self.run()
2025-03-03T16:23:51.4311282Z [DEBUG:LokyProcess-41:MainThread] Worker started with timeout=None
2025-03-03T16:23:51.4311768Z [DEBUG:LokyProcess-41:MainThread] Enabling faulthandler to report tracebacks on worker crashes.
2025-03-03T16:23:51.4312382Z [DEBUG:LokyProcess-43:MainThread] recreated blocker with handle 140120880979968 and name "/loky-2214-vz_11jyo"
2025-03-03T16:23:51.4313021Z [DEBUG:LokyProcess-43:MainThread] recreated blocker with handle 140120856637440 and name "/loky-2214-jb7x32rn"
2025-03-03T16:23:51.4313655Z [DEBUG:LokyProcess-43:MainThread] recreated blocker with handle 140120846254080 and name "/loky-2214-9n25pawk"
2025-03-03T16:23:51.4314290Z [DEBUG:LokyProcess-43:MainThread] recreated blocker with handle 140120846249984 and name "/loky-2214-e48w3t5b"
2025-03-03T16:23:51.4314922Z [DEBUG:LokyProcess-43:MainThread] recreated blocker with handle 140120804052992 and name "/loky-2214-gut13v_x"
2025-03-03T16:23:51.4315559Z [DEBUG:LokyProcess-43:MainThread] recreated blocker with handle 140120880984064 and name "/loky-2214-z9o76fg3"
2025-03-03T16:23:51.4316197Z [DEBUG:LokyProcess-43:MainThread] recreated blocker with handle 140120804040704 and name "/loky-2214-mner7cmc"
2025-03-03T16:23:51.4316800Z [INFO:LokyProcess-43:MainThread] child process calling self.run()
2025-03-03T16:23:51.4317195Z [DEBUG:LokyProcess-43:MainThread] Worker started with timeout=None
2025-03-03T16:23:51.4317679Z [DEBUG:LokyProcess-43:MainThread] Enabling faulthandler to report tracebacks on worker crashes.
2025-03-03T16:23:51.4318286Z [DEBUG:LokyProcess-42:MainThread] recreated blocker with handle 140120880979968 and name "/loky-2214-vz_11jyo"
2025-03-03T16:23:51.4318929Z [DEBUG:LokyProcess-42:MainThread] recreated blocker with handle 140120856637440 and name "/loky-2214-jb7x32rn"
2025-03-03T16:23:51.4319566Z [DEBUG:LokyProcess-42:MainThread] recreated blocker with handle 140120846254080 and name "/loky-2214-9n25pawk"
2025-03-03T16:23:51.4320315Z [DEBUG:LokyProcess-42:MainThread] recreated blocker with handle 140120846249984 and name "/loky-2214-e48w3t5b"
2025-03-03T16:23:51.4320959Z [DEBUG:LokyProcess-42:MainThread] recreated blocker with handle 140120804052992 and name "/loky-2214-gut13v_x"
2025-03-03T16:23:51.4321591Z [DEBUG:LokyProcess-42:MainThread] recreated blocker with handle 140120880984064 and name "/loky-2214-z9o76fg3"
2025-03-03T16:23:51.4322225Z [DEBUG:LokyProcess-42:MainThread] recreated blocker with handle 140120804044800 and name "/loky-2214-nc10sed3"
2025-03-03T16:23:51.4322746Z [INFO:LokyProcess-42:MainThread] child process calling self.run()
2025-03-03T16:23:51.4323133Z [DEBUG:LokyProcess-42:MainThread] Worker started with timeout=None
2025-03-03T16:23:51.4323724Z [DEBUG:LokyProcess-42:MainThread] Enabling faulthandler to report tracebacks on worker crashes.
2025-03-03T16:23:51.4324337Z [DEBUG:LokyProcess-44:MainThread] recreated blocker with handle 140120880979968 and name "/loky-2214-vz_11jyo"
2025-03-03T16:23:51.4324999Z [DEBUG:LokyProcess-44:MainThread] recreated blocker with handle 140120856637440 and name "/loky-2214-jb7x32rn"
2025-03-03T16:23:51.4325632Z [DEBUG:LokyProcess-44:MainThread] recreated blocker with handle 140120846254080 and name "/loky-2214-9n25pawk"
2025-03-03T16:23:51.4326269Z [DEBUG:LokyProcess-44:MainThread] recreated blocker with handle 140120846249984 and name "/loky-2214-e48w3t5b"
2025-03-03T16:23:51.4326995Z [DEBUG:LokyProcess-44:MainThread] recreated blocker with handle 140120804052992 and name "/loky-2214-gut13v_x"
2025-03-03T16:23:51.4327621Z [DEBUG:LokyProcess-44:MainThread] recreated blocker with handle 140120880984064 and name "/loky-2214-z9o76fg3"
2025-03-03T16:23:51.4328264Z [DEBUG:LokyProcess-44:MainThread] recreated blocker with handle 140120804036608 and name "/loky-2214-hcqnuxgw"
2025-03-03T16:23:51.4328778Z [INFO:LokyProcess-44:MainThread] child process calling self.run()
2025-03-03T16:23:51.4329166Z [DEBUG:LokyProcess-44:MainThread] Worker started with timeout=None
2025-03-03T16:23:51.4329650Z [DEBUG:LokyProcess-44:MainThread] Enabling faulthandler to report tracebacks on worker crashes.
2025-03-03T16:23:51.4330135Z --------------------------- Captured stderr teardown ---------------------------
2025-03-03T16:23:51.4330719Z [DEBUG:MainProcess:MainThread] shutting down executor <loky.process_executor.ProcessPoolExecutor object at 0x7f70475b9750>
```

</details>

Here is the archive of the full log:

[5_Testing (linux-py311, ubuntu-latest, 3.11, true).txt](https://github.com/user-attachments/files/19066992/5_Testing.linux-py311.ubuntu-latest.3.11.true.txt)

As far as I know, this particular deadlock only happened once in #419 and in no other PR so it's difficult to tell if it's caused by the changes in #419 or not.

EDIT: it happened twice in #419. Let's see if we can trigger it against the current `master` branch instead.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,7 @@
 ### 3.5.0 - in development
 
+- Fix a random deadlock caused by a race condition at executor shutdown that
+  was observed on Linux and Windows. (#438)
 
 - Fix detection of the number of physical cores in
   `cpu_count(only_physical_cores=True)` on some Linux systems and recent

--- a/loky/process_executor.py
+++ b/loky/process_executor.py
@@ -305,9 +305,11 @@ class _SafeQueue(Queue):
         pending_work_items=None,
         running_work_items=None,
         thread_wakeup=None,
+        shutdown_lock=None,
         reducers=None,
     ):
         self.thread_wakeup = thread_wakeup
+        self.shutdown_lock = shutdown_lock
         self.pending_work_items = pending_work_items
         self.running_work_items = running_work_items
         super().__init__(max_size, reducers=reducers, ctx=ctx)
@@ -336,7 +338,8 @@ class _SafeQueue(Queue):
             if work_item is not None:
                 work_item.future.set_exception(raised_error)
                 del work_item
-            self.thread_wakeup.wakeup()
+            with self.shutdown_lock:
+                self.thread_wakeup.wakeup()
         else:
             super()._on_queue_feeder_error(e, obj)
 
@@ -1139,6 +1142,7 @@ class ProcessPoolExecutor(Executor):
             pending_work_items=self._pending_work_items,
             running_work_items=self._running_work_items,
             thread_wakeup=self._executor_manager_thread_wakeup,
+            shutdown_lock=self._shutdown_lock,
             reducers=job_reducers,
             ctx=self._context,
         )

--- a/tests/_executor_mixin.py
+++ b/tests/_executor_mixin.py
@@ -4,6 +4,7 @@ import math
 import pytest
 import threading
 from time import sleep
+import random
 
 from loky import TimeoutError, get_reusable_executor
 from loky.backend import get_context
@@ -28,6 +29,10 @@ def initializer_event(event):
     """Initializer that set a global test event for test synchronization"""
     global _test_event
     _test_event = event
+
+    # Inject some randomness in the initialization to reveal race conditions.
+    if random.random() < 0.2:
+        sleep(random.random() * 0.1)  # 0-100ms
 
 
 def _direct_children_with_cmdline(p):

--- a/tests/_test_process_executor.py
+++ b/tests/_test_process_executor.py
@@ -112,10 +112,9 @@ class ExecutorShutdownTest:
 
     def test_shutdown_with_sys_exit_at_pickle(self):
         self.executor.shutdown()
-        for i in range(30):
-            with self.executor_type(max_workers=4) as e:
-                with pytest.raises(PicklingError):
-                    e.submit(id, ExitAtPickle()).result()
+        with self.executor_type(max_workers=4) as e:
+            with pytest.raises(PicklingError):
+                e.submit(id, ExitAtPickle()).result()
 
     def test_interpreter_shutdown(self):
         # Free resources to avoid random timeout in CI

--- a/tests/_test_process_executor.py
+++ b/tests/_test_process_executor.py
@@ -107,14 +107,21 @@ class ExecutorShutdownTest:
 
     def test_shutdown_with_pickle_error(self):
         self.executor.shutdown()
-        with self.executor_type(max_workers=4) as e:
-            e.submit(id, ErrorAtPickle())
+        # Iterate a few times to catch deadlocks/race conditions in the
+        # executor shutdown.
+        for _ in range(5):
+            with self.executor_type(max_workers=4) as e:
+                with pytest.raises(PicklingError):
+                    e.submit(id, ErrorAtPickle()).result()
 
     def test_shutdown_with_sys_exit_at_pickle(self):
         self.executor.shutdown()
-        with self.executor_type(max_workers=4) as e:
-            with pytest.raises(PicklingError):
-                e.submit(id, ExitAtPickle()).result()
+        # Iterate a few times to catch deadlocks/race conditions in the
+        # executor shutdown.
+        for _ in range(5):
+            with self.executor_type(max_workers=4) as e:
+                with pytest.raises(PicklingError):
+                    e.submit(id, ExitAtPickle()).result()
 
     def test_interpreter_shutdown(self):
         # Free resources to avoid random timeout in CI

--- a/tests/_test_process_executor.py
+++ b/tests/_test_process_executor.py
@@ -114,7 +114,8 @@ class ExecutorShutdownTest:
         self.executor.shutdown()
         for i in range(30):
             with self.executor_type(max_workers=4) as e:
-                e.submit(id, ExitAtPickle())
+                with pytest.raises(PicklingError):
+                    e.submit(id, ExitAtPickle()).result()
 
     def test_interpreter_shutdown(self):
         # Free resources to avoid random timeout in CI

--- a/tests/_test_process_executor.py
+++ b/tests/_test_process_executor.py
@@ -112,8 +112,9 @@ class ExecutorShutdownTest:
 
     def test_shutdown_with_sys_exit_at_pickle(self):
         self.executor.shutdown()
-        with self.executor_type(max_workers=4) as e:
-            e.submit(id, ExitAtPickle())
+        for i in range(30):
+            with self.executor_type(max_workers=4) as e:
+                e.submit(id, ExitAtPickle())
 
     def test_interpreter_shutdown(self):
         # Free resources to avoid random timeout in CI

